templates/lib/yarp/node.rb.erb

```diff
@@ -6,13 +6,12 @@ module YARP
     attr_reader :<%= param.name %>
 
     <%- end -%>
-    # def initialize: (<%= (node.params.map { |param| "#{param.name}: #{param.rbs_class}" } + ["start_offset: Integer", "length: Integer"]).join(", ") %>) -> void
-    def initialize(<%= (node.params.map(&:name) + ["start_offset", "length"]).join(", ") %>)
+    # def initialize: (<%= (node.params.map { |param| "#{param.name}: #{param.rbs_class}" } + ["location: Location"]).join(", ") %>) -> void
+    def initialize(<%= (node.params.map(&:name) + ["location"]).join(", ") %>)
       <%- node.params.each do |param| -%>
       @<%= param.name %> = <%= param.name %>
       <%- end -%>
-      @start_offset = start_offset
-      @length = length
+      @location = location
```

Comment:
This means the Location object is created eagerly, which is likely significantly slower when creating Ruby nodes.
Basically it reverts https://github.com/ruby/yarp/pull/928/commits/9c57958604d87eff3641c672bdbf6316349b291c
I would suggest to store `start_offset, length, source` instead, and only create the Location object when it's asked for.

Reply:
BTW I was thinking we should also do this for location fields, to object-inline them, in Java and Ruby, since unlike C structs, these languages don't have a way to object-inline in the syntax, it needs to be done manually as multiple fields in the parent object.

Reply:
So I tried inlining it, and it degraded performance for parse and dump/load. I wonder if tripped some threshold where the number of ivars is too large to be using a smaller representation by adding source?

Here are the benchmarks I found:

## 1000.times { YARP.parse_file("lib/yarp/node.rb") }

No inlining:

```
Rehearsal -----------------------------------------
parse   5.587947   0.104401   5.692348 (  5.694816)
-------------------------------- total: 5.692348sec

            user     system      total        real
parse   5.580172   0.096831   5.677003 (  5.677191)
```

Inlining:

```
Rehearsal -----------------------------------------
parse   5.917879   0.181696   6.099575 (  6.162626)
-------------------------------- total: 6.099575sec

            user     system      total        real
parse   6.205026   0.157306   6.362332 (  6.429058)
```

## 1000.times { YARP.load(source, YARP.dump(source)) }

No inlining:

```
Rehearsal -----------------------------------------
parse  18.519872   0.169504  18.689376 ( 18.703276)
------------------------------- total: 18.689376sec

            user     system      total        real
parse  18.937560   0.276980  19.214540 ( 19.350734)
```

Inlining:

```
Rehearsal -----------------------------------------
parse  20.784706   0.186370  20.971076 ( 20.988403)
------------------------------- total: 20.971076sec

            user     system      total        real
parse  20.899019   0.164351  21.063370 ( 21.068951)
```

If you want to mess around with the patch and maybe we can get it faster, I'd be happy to take another look. Here's the patch I was using:

[inline.patch](https://github.com/ruby/yarp/files/11899117/inline.patch)